swap/chain/persistentqueue.go

```diff
@@ -0,0 +1,121 @@
+	The queue provides no dequeue function. Instead an item must be deleted by its key
+*/
+
+// PersistentQueue represents a queue stored in a state store
+type PersistentQueue struct {
+	store   state.Store   // the store backing this queue
+	prefix  string        // the prefix for the keys for this queue
+	trigger chan struct{} // channel to notify the queue that a new item is available
+	nonce   uint64        // increasing nonce. starts with 0 on every startup
+}
+
+// NewPersistentQueue creates a structure to interact with a queue with the given prefix
+func NewPersistentQueue(store state.Store, prefix string) *PersistentQueue {
+	return &PersistentQueue{
+		store:   store,
+		prefix:  prefix,
+		trigger: make(chan struct{}, 1),
+		nonce:   0,
+	}
+}
+
+// Queue puts the necessary database operations for enqueueing a new item into the supplied batch
+// It returns the generated key and a trigger function which must be called once the batch was successfully written
+// This only returns an error if the encoding fails which is an unrecoverable error
+// A lock must be held and kept until after the trigger function was called or the batch write failed
+func (pq *PersistentQueue) Queue(b *state.StoreBatch, v interface{}) (key string, trigger func(), err error) {
+	// the nonce guarantees keys don't collide if multiple transactions are queued in the same second
+	pq.nonce++
+	key = fmt.Sprintf("%d_%08d", time.Now().Unix(), pq.nonce)
+	if err = b.Put(pq.prefix+key, v); err != nil {
+		return "", nil, err
+	}
+
+	return key, func() {
+		select {
+		case pq.trigger <- struct{}{}:
+		default:
+		}
+	}, nil
+}
+
+// Peek looks at the next item in the queue
+// The error returned is either a decode or an io error
+// A lock must be held when this is called and should be held afterwards to prevent the item from being removed while processing
+func (pq *PersistentQueue) Peek(i interface{}) (key string, exists bool, err error) {
+	err = pq.store.Iterate(pq.prefix, func(k, data []byte) (bool, error) {
+		key = string(k)
+		unmarshaler, ok := i.(encoding.BinaryUnmarshaler)
+		if !ok {
+			return true, json.Unmarshal(data, i)
+		}
+		return true, unmarshaler.UnmarshalBinary(data)
+	})
+	if err != nil {
+		return "", false, err
+	}
+	if key == "" {
+		return "", false, nil
+	}
+	return strings.TrimPrefix(key, pq.prefix), true, nil
+}
+
+// Next looks at the next item in the queue and blocks until an item is available if there is none
+// The error returned is either an decode error, an io error or a cancelled context
+// No lock should not be held when this is called. Only a single call to next may be active at any time
+// If the the key is not "", the value exists, the supplied lock was acquired and must be released by the caller after processing the item
+// The supplied lock should be the same that is used for the other functions
+func (pq *PersistentQueue) Next(ctx context.Context, i interface{}, lock *sync.Mutex) (key string, err error) {
```

Comment:
This locking design was the consequence of an earlier version of the code where the request queue and a notification queue were modified at the same time and I wanted to avoid having to hold three locks of three different objects at the same time as the risk of deadlock seemed high. Anyway that case does not exist anymore in this version, so perhaps a lock can be put back into `pq` again. I will attempt a redesign next week.

Reply:
I would suggest to leave it as is for now. Usage of any of the pq functions without holding the main txqueue lock is always wrong. A lock managed by the pq internally is insufficient as in order to make sure that batches don't overlap or trigger signals are not missed locking is required beyond the scope of single functions. I also tried various approaches to put the locking as part of the batch itself but those only further complicated things.

I think for now we should consider `persistentQueue` (which was unexported, this was never supposed to be used elsewhere anyway) as a helper structure that is exclusively used by the txqueue and therefore have it share its lock. I think we should merge this as is to finish work on this codebase and if necessary consider redesigns when migrating transaction sending to bee.

Reply:
My opinion is that we should not share the lock this way as it creates a code that is hard to maintain. I would not like to approve it in this state.

If we are focusing on bee and will not add new features to the swarm repo, we do not need to merge this PR now. But to leave it for the bee project.

Reply:
Fair enough. I'll leave this PR open for further experimentation. Then we can either still merge this at some point in the future or just continue the redesign on a PR on the bee repo (although I assume it will take a while until we get to tx sending there).